File: snake_world_model/env.py
"""Minimal deterministic 10x10 Snake environment for the world-model project.

No graphics, no ML — just the simulator and a small Gym-like API.
"""

import numpy as np

GRID_SIZE = 10

EMPTY, BODY, HEAD, FOOD = 0, 1, 2, 3
UP, DOWN, LEFT, RIGHT = 0, 1, 2, 3

# Row 0 is the top of the grid, so "up" decreases the row index.
_MOVES = {UP: (-1, 0), DOWN: (1, 0), LEFT: (0, -1), RIGHT: (0, 1)}
_OPPOSITE = {UP: DOWN, DOWN: UP, LEFT: RIGHT, RIGHT: LEFT}


class SnakeEnv:
    """Snake on a 10x10 grid. Seed it for fully deterministic episodes."""

    def reset(self, seed: int | None = None) -> np.ndarray:
        self._rng = np.random.default_rng(seed)
        mid = GRID_SIZE // 2
        # Head first, body trailing to the left; the snake faces right.
        self._snake = [(mid, mid), (mid, mid - 1), (mid, mid - 2)]
        self._direction = RIGHT
        self._place_food()
        return self._build_grid()

    def step(self, action: int) -> tuple[np.ndarray, float, bool, dict]:
        # Reverse turns are ignored; the snake keeps its current direction.
        if action != _OPPOSITE[self._direction]:
            self._direction = action

        dr, dc = _MOVES[self._direction]
        head_r, head_c = self._snake[0]
        new_head = (head_r + dr, head_c + dc)

        # The tail vacates its cell this step, so it never counts as a collision.
        r, c = new_head
        in_bounds = 0 <= r < GRID_SIZE and 0 <= c < GRID_SIZE
        if not in_bounds or new_head in set(self._snake[:-1]):
            return self._build_grid(), -1.0, True, {}

        self._snake.insert(0, new_head)
        if new_head == self._food:
            self._place_food()
            return self._build_grid(), 1.0, False, {}

        self._snake.pop()
        return self._build_grid(), 0.0, False, {}

    def render_ascii(self) -> str:
        obs = self._build_grid()
        # Recover a scalar label per cell from the one-hot channels.
        labels = obs.argmax(axis=-1)  # shape (GRID_SIZE, GRID_SIZE)
        symbols = {EMPTY: ".", BODY: "o", HEAD: "H", FOOD: "*"}
        rows = ["".join(symbols[int(cell)] for cell in row) for row in labels]
        return "\n".join(rows)

    def _build_grid(self) -> np.ndarray:
        """Return a (GRID_SIZE, GRID_SIZE, 4) float32 one-hot tensor.

        Channel order matches the constants: EMPTY=0, BODY=1, HEAD=2, FOOD=3.
        Every cell has exactly one channel set to 1.0.
        """
        obs = np.zeros((GRID_SIZE, GRID_SIZE, 4), dtype=np.float32)
        obs[:, :, EMPTY] = 1.0          # all cells start as empty
        for cell in self._snake[1:]:
            obs[cell][EMPTY] = 0.0
            obs[cell][BODY] = 1.0
        obs[self._food][EMPTY] = 0.0
        obs[self._food][FOOD] = 1.0
        obs[self._snake[0]][EMPTY] = 0.0
        obs[self._snake[0]][HEAD] = 1.0
        return obs

    def _place_food(self) -> None:
        occupied = set(self._snake)
        empty = [(r, c) for r in range(GRID_SIZE) for c in range(GRID_SIZE) if (r, c) not in occupied]
        self._food = empty[int(self._rng.integers(len(empty)))]

File: web/src/lib/snake/env.ts
// TypeScript port of snake_world_model/env.py — same rules, JS-seeded RNG.

export const GRID = 10;
export const EMPTY = 0, BODY = 1, HEAD = 2, FOOD = 3;
export const UP = 0, DOWN = 1, LEFT = 2, RIGHT = 3;

const MOVES: [number, number][] = [[-1, 0], [1, 0], [0, -1], [0, 1]];
const OPPOSITE = [DOWN, UP, RIGHT, LEFT];

export function mulberry32(seed: number): () => number {
	let a = seed >>> 0;
	return () => {
		a |= 0; a = (a + 0x6d2b79f5) | 0;
		let t = Math.imul(a ^ (a >>> 15), 1 | a);
		t = (t + Math.imul(t ^ (t >>> 7), 61 | t)) ^ t;
		return ((t ^ (t >>> 14)) >>> 0) / 4294967296;
	};
}

export type Cell = [number, number];

export class SnakeEnv {
	snake: Cell[] = [];
	food: Cell = [0, 0];
	direction = RIGHT;
	private rng: () => number = Math.random;

	reset(seed: number): void {
		this.rng = mulberry32(seed);
		const mid = GRID / 2;
		this.snake = [[mid, mid], [mid, mid - 1], [mid, mid - 2]];
		this.direction = RIGHT;
		this.placeFood();
	}

	/** Advance one step; returns true if the snake died. */
	step(action: number): boolean {
		if (action !== OPPOSITE[this.direction]) this.direction = action;
		const [dr, dc] = MOVES[this.direction];
		const [hr, hc] = this.snake[0];
		const nr = hr + dr, nc = hc + dc;

		const inBounds = nr >= 0 && nr < GRID && nc >= 0 && nc < GRID;
		const hitsBody = this.snake.slice(0, -1).some(([r, c]) => r === nr && c === nc);
		if (!inBounds || hitsBody) return true;

		this.snake.unshift([nr, nc]);
		if (nr === this.food[0] && nc === this.food[1]) {
			this.placeFood();
			return false;
		}
		this.snake.pop();
		return false;
	}

	/** Per-cell class labels, row-major Uint8Array(100). */
	labels(): Uint8Array {
		const g = new Uint8Array(GRID * GRID);
		for (const [r, c] of this.snake.slice(1)) g[r * GRID + c] = BODY;
		g[this.food[0] * GRID + this.food[1]] = FOOD;
		const [hr, hc] = this.snake[0];
		g[hr * GRID + hc] = HEAD;
		return g;
	}

	private placeFood(): void {
		const occupied = new Set(this.snake.map(([r, c]) => r * GRID + c));
		const empty: number[] = [];
		for (let i = 0; i < GRID * GRID; i++) if (!occupied.has(i)) empty.push(i);
		const idx = empty[Math.floor(this.rng() * empty.length)];
		this.food = [Math.floor(idx / GRID), idx % GRID];
	}
}

/** Same data-collection policy as collect.py: walk toward the food. */
export function greedyAction(snake: Cell[], food: Cell, direction: number): number {
	const [hr, hc] = snake[0];
	const [fr, fc] = food;
	let preferred: number, other: number;
	if (Math.abs(fr - hr) >= Math.abs(fc - hc)) {
		preferred = fr > hr ? DOWN : UP;
		other = fc > hc ? RIGHT : LEFT;
	} else {
		preferred = fc > hc ? RIGHT : LEFT;
		other = fr > hr ? DOWN : UP;
	}
	return preferred !== OPPOSITE[direction] ? preferred : other;
}
